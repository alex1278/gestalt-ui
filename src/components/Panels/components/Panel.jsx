import React, { Component } from 'react';
import PropTypes from 'prop-types';
import styled, { withTheme } from 'styled-components';
import { FontIcon } from 'react-md';
import { Title } from 'components/Typography';

const PanelWrapper = styled.div`
  width: 100%;
  box-shadow: 0 2px 2px 0 rgba(0,0,0,.14), 0 1px 5px 0 rgba(0,0,0,.12), 0 3px 1px -2px rgba(0,0,0,.2);
`;

const Header = styled.div`
  display: flex;
  align-items: center;
  box-sizing: border-box;
  height: 4em;
  background-color: ${props => props.theme.colors['$md-grey-50']};
  border-top: 1px solid ${props => props.theme.colors['$md-grey-200']};
  border-bottom: 1px solid ${props => props.theme.colors['$md-grey-200']};
  font-weight: 700;

  &:hover {
    cursor: pointer;
  }
`;

const Content = styled.div`
  background-color: ${props => props.theme.colors['$md-white']};
  ${props => (props.noPadding ? 'padding: 0' : 'padding: 1em')};
  border-bottom: 1px solid ${props => props.theme.colors['$md-grey-200']};
  width: 100%;
  display: ${props => (props.expanded ? 'block' : 'none')};
  ${props => (props.minHeight && `min-height: ${props.minHeight}`)};
`;

const ExpanderIcon = styled(FontIcon)`
  padding: .5em;
  transition: transform 100ms ease;
  transform: ${props => (props.expanded ? 'rotate(0)' : 'rotate(-90deg)')};
`;

class Panel extends Component {
  static propTypes = {
    title: PropTypes.string.isRequired,
    children: PropTypes.any.isRequired,
    defaultExpanded: PropTypes.bool,
    noPadding: PropTypes.bool,
    minHeight: PropTypes.oneOf([PropTypes.bool, PropTypes.string]),
  };

  static defaultProps = {
    defaultExpanded: true,
    noPadding: false,
    minHeight: false,
  };

  state = { isExpanded: this.props.defaultExpanded };

  toggle = () => this.setState({ isExpanded: !this.state.isExpanded });

  render() {
    return (
      <PanelWrapper>
        <Header onClick={this.toggle}>
          <ExpanderIcon expanded={this.state.isExpanded}>expand_more</ExpanderIcon>
          {this.props.title && <Title>{this.props.title}</Title>}
        </Header>
        <Content
          expanded={this.state.isExpanded}
          noPadding={this.props.noPadding}
          minHeight={this.props.minHeight}
        >
          {this.props.children}
        </Content>
      </PanelWrapper>
    );
  }
}

export default withTheme(Panel);
