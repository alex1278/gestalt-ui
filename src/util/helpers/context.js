/**
 * generateContextEntityState
 * @param {*} params
 */
export function generateContextEntityState(params) {
  const entity = {
    id: params.environmentId || params.workspaceId || null,
    key: null,
  };

  if (params.workspaceId && params.environmentId) {
    entity.key = 'environments';
  }

  if (params.workspaceId && !params.environmentId) {
    entity.key = 'workspaces';
  }

  // if (params.workspaceId && params.environmentId && params.providerId) {
  //   entity.id = params.providerId;
  //   entity.key = 'providers';
  // }

  return entity;
}
