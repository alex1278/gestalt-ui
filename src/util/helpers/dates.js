import moment from 'moment';

export const getNow = () => moment();

export const getAgo = (num = 0, unit = 'day') =>
  moment().subtract(num, unit);

export const isRecentWithin = (timestamp, num, unit) =>
  moment(timestamp).isBetween(getAgo(num, unit), getNow());
