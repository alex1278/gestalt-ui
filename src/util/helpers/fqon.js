// sad hack for parsing url

export default function fqonParent(organization) {
  if (organization.org) {
    const urlParser = document.createElement('a');
    urlParser.href = organization.org.href;
    return urlParser.pathname.replace('/', '').split('/')[0];
  }

  return organization;
}
