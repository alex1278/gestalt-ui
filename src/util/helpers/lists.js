export function insertItem(array, payload) {
  return [...array, payload];
}

export function removeItem(array, payload) {
  const newArray = array.slice();
  newArray.splice(newArray.findIndex(a => a === payload), 1);

  return newArray;
}

export function removeItemById(array, id, keyField = 'id') {
  const newArray = array.slice();
  newArray.splice(newArray.findIndex(a => a[keyField] === id), 1);

  return newArray;
}

export function updateItem(array, payload) {
  return array.map((item) => {
    if (item.id === payload.id) {
      return {
        ...payload,
      };
    }

    // Leave every other item unchanged
    return item;
  });
}
