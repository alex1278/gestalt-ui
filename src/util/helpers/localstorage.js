/**
 * getItem
 * @param {*} key
 */
export const getItem = (key) => {
  try {
    return localStorage.getItem(key);
  } catch (err) {
    return undefined;
  }
};

/**
 * saveItem
 * @param {*} key
 * @param {*} value
 */
export const saveItem = (key, value) => {
  try {
    localStorage.setItem(key, value);
  } catch (err) {
    // ignore write errors
  }

  return undefined;
};

/**
 * removeItem
 * @param {*} key
 * @param {*} value
 */
export const removeItem = (key) => {
  try {
    localStorage.removeItem(key);
  } catch (err) {
    // ignore write errors
  }

  return undefined;
};
