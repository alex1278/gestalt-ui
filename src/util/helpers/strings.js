import { startCase, toLower } from 'lodash';

/**
 * getLastFromSplit
 * Parses and returns the last item in a structure like GrandPareng::Parent::Child
 * @param {String} string
 */
export function getLastFromSplit(string, delimiter = '::') {
  if (string) {
    const split = string.split(delimiter);

    return split[split.length - 1];
  }
  return string;
}

/**
 * truncate
 * @param {String} string
 * @param {Number} at
 * @param {String} trailing
 */
export function truncate(string, at, trailing = '...') {
  if (string) {
    if (string.length < at) {
      return string;
    }

    return `${string.substring(0, at).trim()}${trailing}`;
  }

  return string;
}

/**
 * toTitleCase
 * @param {String} string
 */
export function toTitleCase(string) {
  return startCase(toLower(string));
}


// TODO: Will be rmeoved when we have security config
export const checkIfPassword = field =>
  field && field.length && (field.toUpperCase().includes('PASSWORD') || field.toUpperCase().includes('SECRET') || field.toUpperCase().includes('KEY'));

export default {
  getLastFromSplit,
  truncate,
  toTitleCase,
  checkIfPassword
};

export const removeHostFromURL = (url) => {
  if (url) {
    return url.replace(/^.*\/\/[^/]+:?[0-9]?\//i, '');
  }

  return url;
};

export const isBase64 = (str) => {
  try {
    return btoa(atob(str)) === str;
  } catch (err) {
    return false;
  }
};
