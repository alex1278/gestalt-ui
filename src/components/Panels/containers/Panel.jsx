import React, { PureComponent } from 'react';
import PropTypes from 'prop-types';
import { Card } from 'components/Cards';
import styled, { css } from 'styled-components';
import { DotActivity } from 'components/ProgressIndicators';
import Header from '../components/Header';
import Content from '../components/Content';

const PanelWrapper = styled(({ fill, error, expanded, noShadow, noBackground, ...rest }) => <Card noShadow={noShadow} {...rest} />)`
  position: relative;
  width: 100%;
  ${props => props.fill && props.expanded && 'height: 100%'};
  ${props => props.error && !props.expanded && css`
    box-shadow: 0 0 2px 2px ${props.theme.colors['$md-red-500']};
  `};
`;

class Panel extends PureComponent {
  static propTypes = {
    title: PropTypes.string,
    children: PropTypes.oneOfType([
      PropTypes.arrayOf(PropTypes.node),
      PropTypes.node
    ]),
    defaultExpanded: PropTypes.bool,
    noPadding: PropTypes.bool,
    minHeight: PropTypes.oneOfType([PropTypes.bool, PropTypes.string]),
    maxHeight: PropTypes.oneOfType([PropTypes.bool, PropTypes.string]),
    pending: PropTypes.bool,
    expandable: PropTypes.bool,
    count: PropTypes.number,
    error: PropTypes.bool,
    noShadow: PropTypes.bool,
    icon: PropTypes.oneOfType([PropTypes.func, PropTypes.node, PropTypes.bool]),
    fill: PropTypes.bool,
    disabled: PropTypes.bool,
  };

  static defaultProps = {
    title: null,
    defaultExpanded: true,
    noPadding: false,
    minHeight: false,
    maxHeight: false,
    children: null,
    pending: false,
    expandable: true,
    count: 0,
    error: false,
    noShadow: false,
    icon: false,
    fill: false,
    disabled: false,
  };

  state = { isExpanded: this.props.defaultExpanded };

  toggle = () => this.setState(prevState => ({ isExpanded: !prevState.isExpanded }));

  render() {
    const { title, noPadding, minHeight, maxHeight, pending, children, expandable, count, error, noShadow, icon, fill, disabled } = this.props;
    const { isExpanded } = this.state;

    return (
      <PanelWrapper error={error} noShadow={noShadow} fill={fill} expanded={isExpanded}>
        {!!title &&
        <Header
          title={title}
          onClick={expandable ? this.toggle : null}
          expandable={expandable}
          expanded={isExpanded}
          count={count}
          icon={icon}
          noShadow={noShadow}
        />}
        <Content
          isExpanded={isExpanded}
          noPadding={noPadding}
          minHeight={minHeight}
          maxHeight={maxHeight}
          disabled={disabled}
        >
          {pending ? <DotActivity centered size={2} /> : children}
        </Content>
      </PanelWrapper>
    );
  }
}

export default Panel;
