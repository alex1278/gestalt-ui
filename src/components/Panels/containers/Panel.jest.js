import React from 'react';
import Panel from './Panel';
import { shallowWithTheme, mountWithTheme } from '../../../../test/helpers';

describe('(Panels) Panel', () => {
  it('renders component without exploding', () => {
    const wrapper = shallowWithTheme(
      <Panel title="gazobazorb" />
    );

    expect(wrapper).toHaveLength(1);
  });

  it('renders component with basic props', () => {
    const wrapper = shallowWithTheme(
      <Panel title="gazobazorb" />
    );

    // need to dive 2 levels (themeprovider and component) down to get to our wrapper div style
    expect(wrapper.dive().dive()).toMatchSnapshot();
  });

  it('renders component with minHeight prop', () => {
    const wrapper = shallowWithTheme(
      <Panel title="gazobazorb" minHeight="10em" />
    );

    expect(wrapper.dive().dive()).toMatchSnapshot();
  });

  it('renders component with noPadding prop', () => {
    const wrapper = shallowWithTheme(
      <Panel title="gazobazorb" noPadding />
    );

    expect(wrapper.dive().dive()).toMatchSnapshot();
  });

  it('renders component with children', () => {
    const wrapper = shallowWithTheme(
      <Panel title="gazobazorb"><span>this is a test</span></Panel>
    );

    expect(wrapper.dive().dive()).toMatchSnapshot();
  });

  it('renders component activity status when pending', () => {
    const wrapper = shallowWithTheme(
      <Panel title="gazobazorb" pending><span>this is a test</span></Panel>
    );

    expect(wrapper.dive().dive()).toMatchSnapshot();
  });

  it('toggles isExpanded on correct elements when the Header is clicked', () => {
    const wrapper = mountWithTheme(
      <Panel title="gazobazorb"><span>this is a test</span></Panel>
    );
    const header = wrapper.find('WithTheme(Header)');
    const expanderIcon = () => wrapper.find('WithTheme(ExpanderIcon)').props();
    const content = () => wrapper.find('WithTheme(Content)').props();

    header.simulate('click');

    expect(expanderIcon().isExpanded).toBe(false);
    expect(content().isExpanded).toBe(false);
  });
});
