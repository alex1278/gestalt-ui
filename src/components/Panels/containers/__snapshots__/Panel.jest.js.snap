// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`(Panels) Panel renders component activity status when pending 1`] = `
.c0 {
  position: relative;
  width: 100%;
}

<Component
  className="c0"
  error={false}
  expanded={true}
  fill={false}
  noShadow={false}
>
  <WithTheme(Header)
    count={0}
    expandable={true}
    expanded={true}
    icon={false}
    noShadow={false}
    onClick={[Function]}
    title="gazobazorb"
  />
  <WithTheme(Content)
    disabled={false}
    isExpanded={true}
    minHeight={false}
    noPadding={false}
  >
    <DotActivity
      centered={true}
      primary={false}
      size={2}
      visible={true}
    />
  </WithTheme(Content)>
</Component>
`;

exports[`(Panels) Panel renders component with basic props 1`] = `
.c0 {
  position: relative;
  width: 100%;
}

<Component
  className="c0"
  error={false}
  expanded={true}
  fill={false}
  noShadow={false}
>
  <WithTheme(Header)
    count={0}
    expandable={true}
    expanded={true}
    icon={false}
    noShadow={false}
    onClick={[Function]}
    title="gazobazorb"
  />
  <WithTheme(Content)
    disabled={false}
    isExpanded={true}
    minHeight={false}
    noPadding={false}
  />
</Component>
`;

exports[`(Panels) Panel renders component with children 1`] = `
.c0 {
  position: relative;
  width: 100%;
}

<Component
  className="c0"
  error={false}
  expanded={true}
  fill={false}
  noShadow={false}
>
  <WithTheme(Header)
    count={0}
    expandable={true}
    expanded={true}
    icon={false}
    noShadow={false}
    onClick={[Function]}
    title="gazobazorb"
  />
  <WithTheme(Content)
    disabled={false}
    isExpanded={true}
    minHeight={false}
    noPadding={false}
  >
    <span>
      this is a test
    </span>
  </WithTheme(Content)>
</Component>
`;

exports[`(Panels) Panel renders component with minHeight prop 1`] = `
.c0 {
  position: relative;
  width: 100%;
}

<Component
  className="c0"
  error={false}
  expanded={true}
  fill={false}
  noShadow={false}
>
  <WithTheme(Header)
    count={0}
    expandable={true}
    expanded={true}
    icon={false}
    noShadow={false}
    onClick={[Function]}
    title="gazobazorb"
  />
  <WithTheme(Content)
    disabled={false}
    isExpanded={true}
    minHeight="10em"
    noPadding={false}
  />
</Component>
`;

exports[`(Panels) Panel renders component with noPadding prop 1`] = `
.c0 {
  position: relative;
  width: 100%;
}

<Component
  className="c0"
  error={false}
  expanded={true}
  fill={false}
  noShadow={false}
>
  <WithTheme(Header)
    count={0}
    expandable={true}
    expanded={true}
    icon={false}
    noShadow={false}
    onClick={[Function]}
    title="gazobazorb"
  />
  <WithTheme(Content)
    disabled={false}
    isExpanded={true}
    minHeight={false}
    noPadding={true}
  />
</Component>
`;
